src/materialized/src/telemetry.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Telemetry collection.
+
+use log::{debug, log, Level};
+use serde::Deserialize;
+use tokio::time::{self, Duration};
+use uuid::Uuid;
+
+use ore::retry::Retry;
+
+use crate::BUILD_INFO;
+
+/// Telemetry configuration.
+pub struct Config {
+    /// The domain to report telemetry data to.
+    pub domain: String,
+    /// How often to report telemetry data.
+    pub interval: Duration,
+    /// The ID of the Materialize cluster.
+    pub cluster_id: Uuid,
+    /// A coordinator client.
+    pub coord_client: coord::Client,
+}
+
+/// Runs the telemetry reporting loop.
+///
+/// The loop ticks at the interval specified in `config.interval`. On each turn,
+/// it reports various anonymous metadata about the system to the server running
+/// at `config.domain`. The server responds with the latest available version of
+/// Materialize; if that version is
+pub async fn report_loop(config: Config) {
+    let mut interval = time::interval(config.interval);
+    let mut reported_version = BUILD_INFO.semver_version();
+    loop {
+        interval.tick().await;
+
+        let latest_version = match report_one(&config).await {
+            Ok(latest_version) => latest_version,
+            Err(e) => {
+                debug!("failed to report telemetry: {}", e);
+                continue;
+            }
+        };
+
+        if latest_version > reported_version {
+            // We assume users running development builds are sophisticated, and
+            // may be intentionally not running the latest release, so downgrade
+            // the message from warn to info level.
+            let level = match latest_version.pre.as_slice() {
+                [semver::Identifier::AlphaNumeric(s)] if s == "dev" => Level::Info,
+                _ => Level::Warn,
+            };
+            log!(
+                level,
+                "a new version of materialized is available: {}",
+                latest_version
+            );
+            reported_version = latest_version;
+        }
+    }
+}
+
+/// The query used to gather telemetry data.
+const TELEMETRY_QUERY: &str = "SELECT jsonb_build_object(
```

Comment:
This is so good. More interfaces in SQL!

Reply:
This is great! 👍🏼